runner/query_run.go

```diff
@@ -94,7 +95,7 @@ func run(ctx context.Context, server *state.Server, collectionOpts state.Collect
 			db.ExecContext(ctx, postgres.QueryMarkerSQL+"BEGIN READ ONLY")
 
 			err = db.QueryRowContext(ctx, comment+prefix+query.QueryText).Scan(&result)
-			firstErr := err
+			firstErr = err
 
 			// Run EXPLAIN ANALYZE a second time to get a warm cache result
```

Comment:
In https://github.com/pganalyze/collector/pull/641#issuecomment-2501026138, you mentioned that this is executed even if the firstErr is not nil for the case of first one failed with a timeout. However, even though the cause was a timeout, the second one will fail regardless:

```
pgaweb=# set statement_timeout = 1000;
SET
pgaweb=# begin read only;
BEGIN
pgaweb=*# select pg_sleep(2);
ERROR:  canceling statement due to statement timeout
pgaweb=!# select pg_sleep(0.5);
ERROR:  current transaction is aborted, commands ignored until end of transaction block
```

Meaning this code doesn't quite do what you want it to do. When the firstErr is not nil (for whatever the reason), this second query execution will always be ignored.

Reply:
Good catch. Maybe the query text should contain `BEGIN READ ONLY` and `COMMIT` so each query is in its own transaction.